### PR TITLE
[WIP] Integrate xdoctest

### PR DESCRIPTION
**Pull request by Erotemic**
Working branch for concept discussed in #15646

I'm copy-pasting the content from that discussion for context: 

> I'm also going to use this issue as a place to propose the eventual inclusion of xdoctest (a pip installable library I wrote) in pytorch's test suite. I think there are a lot of problems with Python's built in doctest module, and I've built xdoctest to fix them. I would love for my project to get some exposure and its addition to PyTorch may benefit both projects. Please see the readme for more details on what xdoctest brings to the table over the builtin doctest module: https://github.com/Erotemic/xdoctest
> 
> I came across this small syntax error when working on ensuring xdoctest was compatible with pytorch. It isn't 100% there yet, but I'm working on it. My goal is to ensure that xdoctest is 100% compatible with all of torch's doctest out-of-the-box before writing up the PR. I'm also airing the idea out-loud before I commit too much time into this (or get my hopes up), so I'm attaching this little blurb to a no-brainer-merge PR to (1) demonstrate a little bit of value (because xdoctest flagged this syntax error) and (2) see how its received.
> 
>

### CODE REVIEW

**Comment by suo** [comment]
What are the advantages of using xdoctest? The disadvantage is adding another dependency, which places a maintenance burden on the pytorch team and inflicts another package on our users.

**Comment by ezyang** [comment]
@suo To be fair, this is just for testing, so a user wouldn't have to install xdoctest to use PyTorch. I am less than keen on adding `if __name__ == '__main__':` in all of our modules though; don't want to have to write this boilerplate.

I'm not sure if we currently have doctest hooked up at all with our CI. @Erotemic I think making sure this is working properly first would be a good step, then we can talk about swapping it for xdoctest :)

**Comment by Erotemic** [comment]
@suo xdoctest has several advantages over the builtin doctest. Most of these advantages stem from the fact that it is built on top of a lexer and parser instead of regular expressions like the builtin doctest module is. Yes, it is an additional dependency, but currently the regular doctests don't even run/work. I've written xdoctest so that it allows a much more permissive syntax than regular doctest, so it is much easier to actually write, maintain, and ultimately run the doctests. However, the dependency issue shouldn't be a huge deal because (1) xdoctest is pure-python and has [minimal dependencies](https://github.com/Erotemic/xdoctest/blob/master/requirements.txt) and (2) its pip-installable. I've written more about the differences and advantages in a blog post: https://erotemic.wordpress.com/2018/06/12/extending-python-doctests-xdoctest/

A short list of other reasons why xdoctest is worth adding as a dependency: 
* First class support for google-style docstrings
* Parsing of doctests is handled via static analysis instead of dynamically importing the code. (aka its impossible for the tested library to introduce side effects before the actual tests are run). 
* xdoctest handles correctly handles line numbers and reports which line failed wrt the original file and wrt the doctest itself. The builtin doctest module only does the latter. 
* xdoctest is has a pytest plugin and replaces doctest seemlessly. 
* You don't need to differentiate PS1 and PS2 lines (aka you only need to prefix your doctest with `>>> ` and not a special mix of `>>> ` and `... `.)
* Regular doctest is a REPL. Xdoctest can run in REPL mode, but it uses block-style execution by default, which is more similar to how real code is run.
* xdoctest supports multi-line directives, whereas all doctest directives have to be inline. 
* xdoctest can run doctests in pybind11 code. 

@ezyang The `if __name__ == '__main__'` is an artifact from my workflow and is 100% not necessary. I understand why it might be considered understandable, but note It does explicitly show how to invoke and run all doctests within a particular module, which is a reason I keep it around in my personal projects. I do intend on removing it after development is complete though. 

My goal is to fix the pytorch tests such that they work with xdoctest and doctest as much as possible. However, I've already noticed some places where a test simply wont work with regular doctest, so in those instance I may need to use a special xdoctest directive to make it work. In this case most of the tests would run and pass with doctest and 100% of the doctests would run and pass with xdoctest.

**Comment by Erotemic** [comment]
With the release of xdoctest 0.7.0 it is now possible to configure a global environment for all doctests. 

The command: 
`xdoctest -m ./torch/ --global-exec "from torch import nn\nimport torch.nn.functional as F\nimport torch" --options=+IGNORE_WHITESPACE` is able to run all of the doctests. There are 132 doctests that pass, but 94 are failing mostly due to outdated got/want statements. 

The list of these failing doctests is as follows:

```
python -m xdoctest torch set_default_dtype:0
python -m xdoctest torch.serialization load:0
python -m xdoctest torch.functional einsum:0
python -m xdoctest torch.functional argmax:0
python -m xdoctest torch.functional argmin:0
python -m xdoctest torch.functional tensordot:0
python -m xdoctest torch.functional argsort:0
python -m xdoctest torch.functional cartesian_prod:0
python -m xdoctest torch.functional norm:0
python -m xdoctest torch.functional chain_matmul:0
python -m xdoctest torch.sparse mm:0
python -m xdoctest torch.sparse sum:0
python -m xdoctest torch.jit load:0
python -m xdoctest torch.jit save:0
python -m xdoctest torch.jit get_trace_graph:0
python -m xdoctest torch.nn.grad conv1d_weight:0
python -m xdoctest torch.nn.grad conv2d_weight:0
python -m xdoctest torch.nn.functional fractional_max_pool3d_with_indices:0
python -m xdoctest torch.nn.functional embedding:0
python -m xdoctest torch.nn.functional embedding_bag:0
python -m xdoctest torch.nn.modules.padding ConstantPad1d:0
python -m xdoctest torch.nn.modules.padding ConstantPad2d:0
python -m xdoctest torch.nn.modules.padding ReflectionPad1d:0
python -m xdoctest torch.nn.modules.padding ReflectionPad2d:0
python -m xdoctest torch.nn.modules.padding ReplicationPad1d:0
python -m xdoctest torch.nn.modules.padding ReplicationPad2d:0
python -m xdoctest torch.nn.modules.padding ZeroPad2d:0
python -m xdoctest torch.nn.modules.upsampling Upsample:0
python -m xdoctest torch.nn.modules.upsampling UpsamplingNearest2d:0
python -m xdoctest torch.nn.modules.upsampling UpsamplingBilinear2d:0
python -m xdoctest torch.nn.modules.sparse Embedding.from_pretrained:0
python -m xdoctest torch.nn.modules.sparse EmbeddingBag.from_pretrained:0
python -m xdoctest torch.nn.modules.fold Fold:0
python -m xdoctest torch.nn.modules.fold Unfold:0
python -m xdoctest torch.nn.modules.pooling MaxUnpool1d:0
python -m xdoctest torch.nn.modules.pooling FractionalMaxPool3d:0
python -m xdoctest torch.nn.modules.module Module.register_buffer:0
python -m xdoctest torch.nn.modules.module Module.apply:0
python -m xdoctest torch.nn.modules.module Module.to:0
python -m xdoctest torch.nn.modules.module Module.state_dict:0
python -m xdoctest torch.nn.modules.module Module.parameters:0
python -m xdoctest torch.nn.modules.module Module.named_parameters:0
python -m xdoctest torch.nn.modules.module Module.buffers:0
python -m xdoctest torch.nn.modules.module Module.named_buffers:0
python -m xdoctest torch.nn.modules.module Module.named_children:0
python -m xdoctest torch.nn.modules.module Module.named_modules:0
python -m xdoctest torch.nn.parallel.distributed_cpu DistributedDataParallelCPU:0
python -m xdoctest torch.nn.parallel.data_parallel DataParallel:0
python -m xdoctest torch.nn.parallel.distributed DistributedDataParallel:0
python -m xdoctest torch.nn.parallel.deprecated.distributed_cpu DistributedDataParallelCPU:0
python -m xdoctest torch.nn.parallel.deprecated.distributed DistributedDataParallel:0
python -m xdoctest torch.nn.utils.spectral_norm spectral_norm:0
python -m xdoctest torch.nn.utils.weight_norm weight_norm:0
python -m xdoctest torch.autograd.grad_mode no_grad:0
python -m xdoctest torch.autograd.grad_mode enable_grad:0
python -m xdoctest torch.autograd.grad_mode set_grad_enabled:0
python -m xdoctest torch.autograd.profiler profile:0
python -m xdoctest torch.autograd.profiler emit_nvtx:0
python -m xdoctest torch.utils.checkpoint checkpoint_sequential:0
python -m xdoctest torch.utils.cpp_extension load:0
python -m xdoctest torch.optim.lr_scheduler LambdaLR:0
python -m xdoctest torch.optim.lr_scheduler StepLR:0
python -m xdoctest torch.optim.lr_scheduler MultiStepLR:0
python -m xdoctest torch.optim.lr_scheduler ReduceLROnPlateau:0
python -m xdoctest torch.optim.sgd SGD:0
python -m xdoctest torch.distributions.lowrank_multivariate_normal LowRankMultivariateNormal:0
python -m xdoctest torch.distributions.chi2 Chi2:0
python -m xdoctest torch.distributions.normal Normal:0
python -m xdoctest torch.distributions.logistic_normal LogisticNormal:0
python -m xdoctest torch.distributions.fishersnedecor FisherSnedecor:0
python -m xdoctest torch.distributions.laplace Laplace:0
python -m xdoctest torch.distributions.multivariate_normal MultivariateNormal:0
python -m xdoctest torch.distributions.binomial Binomial:0
python -m xdoctest torch.distributions.beta Beta:0
python -m xdoctest torch.distributions.poisson Poisson:0
python -m xdoctest torch.distributions.independent Independent:0
python -m xdoctest torch.distributions.relaxed_bernoulli RelaxedBernoulli:0
python -m xdoctest torch.distributions.log_normal LogNormal:0
python -m xdoctest torch.distributions.bernoulli Bernoulli:0
python -m xdoctest torch.distributions.half_normal HalfNormal:0
python -m xdoctest torch.distributions.multinomial Multinomial:0
python -m xdoctest torch.distributions.exponential Exponential:0
python -m xdoctest torch.distributions.uniform Uniform:0
python -m xdoctest torch.distributions.weibull Weibull:0
python -m xdoctest torch.distributions.gumbel Gumbel:0
python -m xdoctest torch.distributions.studentT StudentT:0
python -m xdoctest torch.distributions.relaxed_categorical RelaxedOneHotCategorical:0
python -m xdoctest torch.distributions.gamma Gamma:0
python -m xdoctest torch.distributions.half_cauchy HalfCauchy:0
python -m xdoctest torch.distributions.cauchy Cauchy:0
python -m xdoctest torch.distributions.categorical Categorical:0
python -m xdoctest torch.distributions.one_hot_categorical OneHotCategorical:0
python -m xdoctest torch.distributions.pareto Pareto:0
python -m xdoctest torch.distributions.dirichlet Dirichlet:0
```

I plan on working my way through these over the next few days to fix all the easy ones.

**Comment by Erotemic** [comment]
Quick update: 

I've currently got pytorch building in a docker environment, so I'm no longer blocked on that front. 

Current status of doctest updates is that 3 are failing and 223 are passing with 7 total warnings. So, I'm very close to having this branch ready for review. After I fix the tests I just need to add xdoctest to the CI test runner. 

For details the current failing tests are:

```
python -m xdoctest torch.functional unique:0
python -m xdoctest torch.sparse sum:0
python -m xdoctest torch.nn.modules.module Module.named_parameters:0
```

I will try to fix these soon.

**Comment by Erotemic** [comment]
I've rebased on the current master and run the doctests and all 226 doctests now pass! This not only enhances the testing suite, but it ensure that the examples in the docstrings are up to date. 

So the last bit needed is to ensure that the doctests run in the CI. The pytorch CI setup seems very complicated, so I may need a little guidance on exactly where things need to be hooked up. 
For now I've:
* included a script `run_doctests.sh` in the test directory that simply invokes the doctests and
* Modified `test/run_test.py` to enable `xdoctest` when the pytest runner is being used. 

What I don't see is where the requirement for pytest is specified as I would also need to require the xdoctest plugin as well to ensure the CI grabs the package. 

In short: 

* Where should I encode the requirement for xdoctest?
* Where should I place the code that either (a) calls xdoctest's standalone runner or (b) calls pytest with the xdoctest plugin enabled?

**Comment by Erotemic** [comment]
I think the torch devs should consider using xdoctest. It adds quite a bit more flexibility to the type of docstrings that you can write. But torch is so huge now I understand that the dependency burden is an issue.

But currently I don't think you are executing your doctests in your testing suite (at least torch wasn't when I started this PR). So if you do want your doctests to run in your CI, let me know.

File: torch/distributions/binomial.py [hunk 1]
```diff
@@ -14,13 +14,13 @@ class Binomial(Distribution):
     Example::
 
         >>> m = Binomial(100, torch.tensor([0 , .2, .8, 1]))
-        >>> x = m.sample()
-        tensor([   0.,   22.,   71.,  100.])
+        >>> m.sample()
+        tensor([..., ..., ..., ...])
```
Comment (ezyang) [inline torch/distributions/binomial.py:18]: this change is surprising

**Comment by ezyang** [comment]
This is my bad, I missed your last update, I wish I had seen it :(

Let's add xdoctest. For ease of reviewing, I'll suggest doing a split: let's do a PR to get xdoctest hooked into the CI but ignore all failing test cases, and then chunk up PRs fixing various issues. We have direct merge to GitHub now so it should be easier to get diff flow going. I will sponsor reviews and merges of the changes.

File: torch/distributions/binomial.py (hunk 1 above)
Comment (Erotemic) [reply torch/distributions/binomial.py:18]: That's because there is no way to do local random seeding. The output is random so we can't match it.

If I was doing this in numpy I'd do:

```
>>> import numpy as np
>>> rng = np.random.RandomState(0)
>>> rng.randint(0, 10, 5)
array([5, 0, 3, 3, 7])
```

and I wouldn't have to worry that the output would fail. Here I used the ellipsis matching inherited from the builtin doctest to match anything, but this either would need to be fixed by seeding the RNG or disabling output checking (which is easy with `# xdoctest: +IGNORE_WANT` - also not I'm open to adding an alias for that directive, I don't think `IGNORE_WANT` is the best name to indicate we aren't going to check the stdout / output.)

Comment (ezyang) [reply torch/distributions/binomial.py:18]: Global random seeding (hidden from the docs) seems pretty reasonable to make the test work out. Although using the explicit RNG API would also work, we do have an obligation to show users the non-explicit rng state api as well in the docs.

**Comment by Erotemic** [comment]
> This is my bad, I missed your last update, I wish I had seen it :(
> 
> Let's add xdoctest. For ease of reviewing, I'll suggest doing a split: let's do a PR to get xdoctest hooked into the CI but ignore all failing test cases, and then chunk up PRs fixing various issues. We have direct merge to GitHub now so it should be easier to get diff flow going. I will sponsor reviews and merges of the changes.

That sounds good. I'll have to find some time to actually make it happen. I have a few other priorities I need to take care of, so it might take some time.

**Comment by Erotemic** [comment]
This is getting close to the top on my list of FOSS things to do. It's been awhile, but I will likely get to this by the end of the summer.

**Comment by Erotemic** [comment]
@ezyang I'm closing this in favor of #82797
